src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Generators
+{
+    [Generator]
+    public partial class EventSourceGenerator : ISourceGenerator
+    {
+        public void Initialize(GeneratorInitializationContext context)
+            => context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            var receiver = context.SyntaxReceiver as SyntaxReceiver;
+            if (receiver is null || receiver.CandidateClasses is null || receiver.CandidateClasses.Count == 0)
+            {
+                // nothing to do yet
+                return;
+            }
+
+            var p = new Parser(context.Compilation, context.ReportDiagnostic, context.CancellationToken);
+            var eventSources = p.GetEventSourceClasses(receiver.CandidateClasses);
+
+            if (eventSources?.Length > 0)
+            {
+                var e = new Emitter(context);
+                e.Emit(eventSources, context.CancellationToken);
+            }
+        }
+
+        private sealed class SyntaxReceiver : ISyntaxReceiver
+        {
+            private List<ClassDeclarationSyntax>? _candidateClasses;
+
+            public List<ClassDeclarationSyntax>? CandidateClasses => _candidateClasses;
+
+            public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
+            {
+                // Only add classes annotated [EventSourceAutoGenerate] to reduce busy work.
+                const string EventSourceAttribute = "EventSourceAutoGenerateAttribute";
```

Comment:
This won't handle Aliases like `using MyAttrib = System.EventSourceAutoGenerateAttribute;` 

Given the usage only in the repo it might be ok, but something to consider.

Reply:
Wanted a fast bail without expanding the syntax tree as there are a lot of attributed classes in corelib.

Expand the syntax in the parser to compare to actual attribute; though obvs it would be too late then as its already discounted.